logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageSender.kt

```diff
@@ -78,28 +91,34 @@ class MessageSenderImpl(
                 }
         }
 
-    private suspend fun attemptToSendWithMLS(conversationId: ConversationId, groupId: String, message: Message): Either<CoreFailure, Unit> = suspending {
+    private suspend fun attemptToSendWithMLS(
+        conversationId: ConversationId,
+        groupId: String,
+        message: Message
+    ): Either<CoreFailure, String> = suspending {
         mlsMessageCreator.createOutgoingMLSMessage(groupId, message).flatMap { mlsMessage ->
             // TODO handle mls-stale-message
-            messageRepository.sendMLSMessage(conversationId, mlsMessage)
+            messageRepository.sendMLSMessage(conversationId, mlsMessage).map {
+                message.date //TODO return actual server time from the response
+            }
         }
     }
 
     private suspend fun trySendingEnvelopeRetryingIfPossible(
         conversationId: ConversationId,
         envelope: MessageEnvelope,
         messageUuid: Message,
-    ): Either<CoreFailure, Unit> = suspending {
+    ): Either<CoreFailure, String> = suspending {
         messageRepository.sendEnvelope(conversationId, envelope).coFold(
             {
                 when (it) {
                     is SendMessageFailure.Unknown -> Either.Left(it)
                     is SendMessageFailure.ClientsHaveChanged -> messageSendFailureHandler.handleClientsHaveChangedFailure(it).flatMap {
-                        trySendingOutgoingMessage(conversationId, messageUuid)
+                        attemptToSend(conversationId, messageUuid)
```

Comment:
Does this mean that in case of a retry the message time won't be updated?

Reply:
It has been implemented so that there is a loop when `ClientsHaveChanged` (`attemptToSend` -> `attemptToSendWithProteus` -> `trySendingEnvelopeRetryingIfPossible` -> `attemptToSend` ...) and I don't know if it should be this way or not so I didn't want to modify it, just apply the "update message time" changes. It should eventually return the message time when coming out of this loop when `messageRepository.sendEnvelope` returns success.

Reply:
Oh yeah indeed.

We probably should have a `maxRetries` and `retryCount` for this Proteus loop in the future. But it should just work after one or MAX two tries.

Anything other than that would be a very unexpected behaviour from the backend or some internal bug in Kalium.